### PR TITLE
Added Set/Get Chef Extension commands

### PR DESCRIPTION
**Pull request by siddheshwar-more**
Implemented following chef extension related commands-
1. azure vm extension set-chef
2. azure vm extension get-chef

set-chef - Adds the chef extension to VM
get-chef - Gives chef extension details  

Thank you!

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@siddheshwar-more__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by Microsoft and real humans are currently evaluating your PR.</p>

TTYL, AZPRBOT;

File: lib/commands/asm/vm._js [hunk 1]
```diff
@@ -177,6 +177,23 @@ exports.init = function(cli) {
...
 
+  extension.command('set-chef <vm-name>')
+    .description($('Enable/disable Chef extensions for VMs'))
+    .option('-V, --version <number>', $('extension\'s version number. Default is latest.'))
+    .option('-r, --reference-name <name>', $('extension\'s reference name'))
+    .option('-R, --run-list <run-list>', $('Runlist of roles/recipes to apply to VM'))
+    .option('-O, --validation-pem <validation-pem>', $('chef validation pem file path'))
+    .option('-c, --client-config <client-config>', $('chef client configuration file(i.e client.rb) path'))
+    .option('-b, --auto-update-client', $('auto update chef client'))
+    .option('-b, --disable', $('disable extension'))
```
Comment (amarzavery) [inline lib/commands/asm/vm._js]: you cannot have same short hand switch names for two parameters

File: lib/commands/asm/vm/vmclient.js [hunk 2]
```diff
@@ -2743,34 +2844,39 @@ _.extend(VMClient.prototype, {
...
+      showExtensionDetails(role, options, logger, self.cli);
+
+      return callback();
+    });
+  },
+
+  getChefExtension: function(vmName, options, callback, logger) {
+    var self = this;
+
+    options.publisherName = 'Chef.Bootstrap.WindowsAzure';
```
Comment (amarzavery) [inline lib/commands/asm/vm/vmclient.js]: Will the publisher name always be 'Chef.Bootstrap.WindowsAzure' ? If so, how about making it a constant?

File: lib/commands/asm/vm/vmclient.js [hunk 3]
```diff
@@ -2690,38 +2679,150 @@ _.extend(VMClient.prototype, {
...
+    var self = this;
+
+    // if there's no validation.pem and client-config are mandatory
+    if (!options.validationPem || !options.clientConfig) {
+      return callback(
+        new Error($('Required --validation-pem and --client-config options.')));
+    }
+
+    // Set defaults
+    var publisherName = 'Chef.Bootstrap.WindowsAzure';
```
Comment (amarzavery) [inline lib/commands/asm/vm/vmclient.js]: Making this a constant at the top of the file would be a better idea.

**Comment by amarzavery** [comment]
You need to record the tests while running them live. This will create a recordings file in the test\recordings directory. 

Please have a look at this wiki page on how to record tests and which envt. variables you need to set.
https://github.com/Azure/azure-xplat-cli/wiki/Setting-up-environment-variables-for-running-xplat-cli-tests

**Comment by siddheshwar-more** [comment]
@amarzavery 

I've set all environment variable required to run tests.
But still some `cli affinity-group` related tests are failing on my local as well as on Travis.

I can see same tests are failed for #1560 as well! Refer - https://travis-ci.org/Azure/azure-xplat-cli/jobs/49051456#L4127 

I guess its the common issue. If you know the solution, could you please share the details?

Actually my changes are related to extension commands. I ran only extension related tests by commenting other tests from ** tests/testlist.txt** file and all extension tests are passed.

Thank you.

**Comment by yugangw-msft** [comment]
I still don't see the recording files. When you run live tests in record mode then you will see recording files in the test/recordings directory. 
You have to check them in as well as a part of the PR.

Do not worry about the tests that are not a part of your work

**Comment by amarzavery** [comment]
Some CLI Tests will fail for now as we have had a breaking change in Node SDK and have not yet published a new version of node sdk where those changes have been fixed.

**Comment by amarzavery** [comment]
```
set AZURE_NOCK_RECORD=true
```
Comment out other tests in testlist.txt
Only run your tests in Record Mode.
You will see recorded files in test\recordings directory.
The recorded files willl have the same name as ur test files.

To verify that recording was successfull

set envt variable 
```
set AZURE_NOCK_RECORD=
set NOCK_OFF=
```

and run your recorded tests in mock mode..
Your tests should run successfully..

Everything looks good then commit those files in to the PR.

**Comment by siddheshwar-more** [comment]
@amarzavery
I've set-up all test environment and also commented other tests. But I got following response during set chef extension unit tests run.

https://gist.github.com/siddheshwar-more/8d6eb4eccec0115fe3f1 

I've compare this ^ file with existing extension recording files. In above 
https://gist.github.com/siddheshwar-more/8d6eb4eccec0115fe3f1#file-cli-vm-extension_set-chef-nock-js-L27 is empty?

Can you please help us in this matter?

Thank you!

File: test/data/set-chef-extension-validation.pem [hunk 4]
```diff
@@ -0,0 +1,3 @@
...
...Mco9pTY+avM0azIu8pa7Rd5RLiFgZB03N5mTRFfzgLAxFS2dPIdGHJ9KRxmvGTy/xGWd+Jt4f24fv+457J1Fy9ORSZu5M7Q9I0G+Gl+lNfS3x/QUw9ahoqf5SWaFaqUPFZPvQtWbp1nxJQt1PejLStYyF5SnGung4OHNcrFLaE+mKmV3CTV1ZIlbGN6GYKQeU27V7jthis-is-really-fakecTrsk6jZLz3m9MtDqrbN9rklnFJPMM9K4KTkLt8k6VJoiB66tNmNdozhjTId/V033mkQUPj0vTf0qQpiLeEnnBplb2Ljtc9gi6UbGAxZpMCgYEA5BrzIcEU+Pg3KL8Yc5DaXiqPPoRL4+sXqJ30guJlWxsZc4U7GRmbS6xND9lU7ubcxdVPMcTYQMKiDep5tcvuDJ1auP+gFs3J+P3bPJbc9rU8jrPmOuKfey7fmUGlf4eMXh+3mx+8znX6+Wl23RT3skFGo2S2LesjTMdolVo1NB8CgYEAy4zKgNF9FYIGZGiydq2y+lE4jkeSHhlucPFAxoW/Qv8+SKqPAzto30LpfAfnTfRN7xnbGsqpEJtdg59FD8EOuDG7541slajylTHWZpGhQ0NJS/Bhcsld8IIFFUMZoq8QCNJdfSqPwjqq6C4xhF4rya2ZZQsfkgtLwnfRzAxkWX8CfwnNHqVUppIMGFKn42UT1E43hfApvxkMAPg90UPOdoxznaRZFTD+6K1Xuff0XrHlOZdGkyCQVLpzooL2kEQ/fvsLx9orIbXZIdSRE9qGoPf40aFQGZRuKGwil3j+WR4htOZWdo0zz+kPKLmOzoyxWfNO2roqmalfcSTbbKL+PUsCgYEAlNibrJ+ETJoMephlCIv0eT+zj7eejgwiysxghcuAFUn5DcJLWh2zjq+rdVuCNWsVImdR5cGU5+P0+i8ZnW+bh+42bhjo8nfdDGSpV8/ekQU/MBzT5dCCtUEe3nF26ZQCsbDDVKgu/+VO3QIKi7UhaBVeOplruKwx9j8QV5ooN5kCgYB2+L/HG3G4s/oBsUMQI64niYujP0wWi3m8fwxlZWyVLBRjOzWyt9Scge35lsbV+U5/Z/adVSJErH+7tgkCdS/Ibwztz5TvrRFeGIq+dCpDS+Tdo5AsJziuHKAFPfQPzsqRlsb5G/VNmeQd1oY7nQXLWAfcAtIrQbHORj4U4e/Izg==
```
Comment (amarzavery) [inline test/data/set-chef-extension-validation.pem]: Is this a dummy cert or an actual cert?

**Comment by amarzavery** [comment]
@siddheshwar-more 

I am sure you would have done this, however I want to verify:

You should only set the envt variable for nock recording:
```
set AZURE_NOCK_RECORD=true
```

The envt variable to run tests live should be unset while recording tests:
```
set NOCK_OFF=
```

If this is not done, it may interfere with recording tests.

File: test/data/set-chef-extension-validation.pem (hunk 4 above)
Comment (siddheshwar-more) [inline test/data/set-chef-extension-validation.pem]: Its a dummy cert !

File: test/commands/cli.vm.extension_get-chef-tests.js [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    function getVM(callback) {
+      if (getVM.VMName) {
+        callback(getVM.VMName);
+      } else {
+        var cmd = util.format('vm list --json').split(' ');
+        testUtils.executeCommand(suite, retry, cmd, function(result) {
+          result.exitStatus.should.equal(0);
+          var vmList = JSON.parse(result.text);
+          var found = vmList.some(function(vm) {
+            if (vm.OSDisk.operatingSystem.toLowerCase() === 'windows') {
```
Comment (amarzavery) [inline test/commands/cli.vm.extension_get-chef-tests.js]: What if the subscription does not have any provisioned windows vm ? The test should be robust in creating one if it cannot find one and delete the vm after the test is over.

**Comment by amarzavery** [comment]
I was able to record the get-chef tests successfully. I am having a problem running the set-chef tests.

I get the error:
 cli vm extension: Set Chef extensions: error: A VM with dns prefix "xplattestvm" already exists

Reason being there is no clean up (deleting the vm ) done in after each. 

Either do the clean up in after each or make the before each smarter. Create a vm with the name xplattestvm if you cannot fin one. Otherwise use the existing one.


For recording tests:
1. Login using an org id or microsoft.com account (azure login -u$username -p $password)
2. Set the azure subscription under which the testing would be done (azure account set "Subscription name"
3. set NOCK_OFF=
4. set AZURE_NOCK_RECORD=true
5. Create recording files in test/recordings directory if they do not exist
6. Comment out other tests and run your tests in testlist.txt one by one to record.
7. Once tests are recorded, run them mocked. For that set the envt. variables:
set NOCK_OFF=
set AZURE_NOCK_RECORD=

**Comment by siddheshwar-more** [comment]
@amarzavery -
Thank you. Now I'm able to record live traffic , working on review comments!

Live traffic has sensitive data related to azure account, I'm planning to replace that with existing extension related test like- https://github.com/Azure/azure-xplat-cli/blob/master/test/recordings/cli.vm.extension-tests.nock.js#L9 
is it fine?

**Comment by amarzavery** [comment]
@siddheshwar-more 
That should be fine. The subscription id should not be a problem. It is just a guid. Making sure that recorded file does not have cert or access token is a good idea.

**Comment by siddheshwar-more** [comment]
@amarzavery -
I've fixed all review comments related to tests. Also added test recording files and did testing with live and mock mode. All tests are passed.

Could you please merge this PR?

Thank you!

**Comment by amarzavery** [comment]
Awesome! This looks good and the CI is passing as well. Good job :+1:

**Pull request merged**